include/wlr/interfaces/wlr_keyboard.h

```diff
@@ -7,11 +7,16 @@
 struct wlr_keyboard_impl {
 	void (*destroy)(struct wlr_keyboard *keyboard);
 	void (*led_update)(struct wlr_keyboard *keyboard, uint32_t leds);
+	void (*update_state)(struct wlr_keyboard *keyboard,
+			struct wlr_event_keyboard_key *event);
```

Comment:
Maybe we could add a `bool sends_modifiers_update` field instead? Not sure if it's a good idea to put a `bool` in a `*_impl` struct.

Reply:
Nah, I think better solution will be `source` field in `struct wlr_event_keyboard_key` and condition on it.

Reply:
You mean, a field which indicates if keyboard state has already been updated?

Reply:
Something like
```cpp
 void wlr_keyboard_update_state(struct wlr_keyboard *keyboard,
  		struct wlr_event_keyboard_key *event) {
	if (event->source == LIBINPUT) {
		uint32_t keycode = event->keycode + 8;
		xkb_state_update_key(keyboard->xkb_state, keycode,
			event->state == WLR_KEY_PRESSED ? XKB_KEY_DOWN : XKB_KEY_UP);
	}
```

Reply:
libinput may or may not be the only backend that doesn't call `wlr_keyboard_update_modifiers`. I would prefer a more descriptive name, such as `bool update_state`. And as soon as the event will be forwarded to the compositor and the compositor doesn't care about this, maybe make this an additional argument to this function. But you're right, it's a better idea.

Reply:
-1 to having wlr_keyboard test for which backed the event came from